processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java

```diff
@@ -346,4 +349,99 @@ public void close()
     }
     aggBuffers.clear();
   }
+
+  public static class Builder extends AppendableIndexBuilder
+  {
+    @Nullable
+    NonBlockingPool<ByteBuffer> bufferPool = null;
+
+    public Builder setBufferPool(final NonBlockingPool<ByteBuffer> bufferPool)
+    {
+      this.bufferPool = bufferPool;
+      return this;
+    }
+
+    @Override
+    public void validate()
+    {
+      super.validate();
+      if (bufferPool == null) {
+        throw new IllegalArgumentException("bufferPool cannot be null");
+      }
+    }
+
+    @Override
+    protected OffheapIncrementalIndex buildInner()
+    {
+      return new OffheapIncrementalIndex(
+          Objects.requireNonNull(incrementalIndexSchema, "incrementalIndexSchema is null"),
+          deserializeComplexMetrics,
+          concurrentEventAdd,
+          sortFacts,
+          maxRowCount,
+          Objects.requireNonNull(bufferPool, "bufferPool is null")
+      );
+    }
+  }
+
+  public static class Spec implements AppendableIndexSpec, Supplier<ByteBuffer>
+  {
+    public static final String TYPE = "offheap";
+    static final int DEFAULT_BUFFER_SIZE = 1 << 23;
+    static final int DEFAULT_CACHE_SIZE = 1 << 30;
+
+    final int bufferSize;
+    final int cacheSize;
+    final NonBlockingPool<ByteBuffer> bufferPool;
+
+    @JsonCreator
+    public Spec(
+        final @JsonProperty("bufferSize") @Nullable Integer bufferSize,
+        final @JsonProperty("cacheSize") @Nullable Integer cacheSize
+    )
+    {
+      this.bufferSize = bufferSize != null && bufferSize > 0 ? bufferSize : DEFAULT_BUFFER_SIZE;
+      this.cacheSize = cacheSize != null && cacheSize > this.bufferSize ? cacheSize : DEFAULT_CACHE_SIZE;
+      this.bufferPool = new StupidPool<>(
+          "Offheap incremental-index buffer pool",
+          this,
+          0,
+          this.cacheSize / this.bufferSize
```

Comment:
The changes in this class apart from conforming to the AppendableIndexSpec interface might be out of scope of this PR and it might require its own additional tests. Would it be make sense to propose these changes in an incremental PR once the AppendableIndexSpec changes are merged?

Reply:
Yes. That is possible. I will separate this part into a PR that will follow this one.

Reply:
Removed